### ISSUE TITLE
Single version backward compatible SPI

### ISSUE THREAD

**Issue by kokosing** [trinodb/trino#11667]
The goal of this issue is to **guarantee that plugin that is using non deprecated Trino SPI in version X can be used also in Trino in version X+1**.

The Trino SPI is amazing. Trino have so many great plugins that allows connectivity with different services. Moreover users can write their own plugins to go beyond that.

Currently SPI is versioned with the Trino version. There is a best effort convention to not introduce breaking changes, however this convention is not enforced and so breaking changes happen from time to time.

Providing a guarantee that plugin that is using non deprecated Trino SPI in version X can be used also in Trino in version X+1 gives: 
 - possibility decouple update Trino service without a need to update used custom plugins used in previous version (as long no deprecated SPI is used). Update plugins code can be done as separate step. That makes update of Trino way easier.
 - certainty to custom Trino plugin developers what to expect, that makes their life easer in regards to custom Trino plugin maintenance
 
To make this happen we need to prove that non of SPI got changed accidentally. Like no non deprecated methods got removed or changed. Hence I suggest to have a test that would be executed in every PR to verify that non deprecated API is still available, by using Java reflections or other bytecode tools.
 
There is also a possibility that SPI is there but its semantic changed. I do not see a way to test this automatically. However this kind of changes are really bad practices and I believe that code review process would prevent them.

**Comment by martint** [comment]
>  breaking changes happen from time to time.

Unfortunately, sometimes they are necessary. I don't think we can enforce non-breaking changes without a way to allow for breaking changes when we need them.

**Comment by kokosing** [comment]
You are right. However, having an enforcer for that would prevent us from making such changes accidentally. We can always make exception when needed, but they would be deliberate.

**Comment by findepi** [comment]
if this is a blocker, then maybe we have a template, or exclusions support from the start in the test, so that we don't need to invent it when a warranted breakage is being introduced?

such support should also force cleanup of exclusions when we cut a new release (auto reset), but that would also mean project state becomes red after  a release is cut.

cc @losipiuk @nineinchnick

**Comment by kokosing** [comment]
> What would be the mechanism for such exceptions?

See https://github.com/trinodb/trino/pull/11722/files#diff-24d9cf3d6431ec3dff2f35927e19c6f9474dc2b1c9c7de86ab031644a999549cR40

**Comment by electrum** [comment]
Should we close this now that #11722 is merged? I don't see anything actionable left here.
